### PR TITLE
chore(RHTAPREL-304): add missing metrics for SLO alerts

### PR DESCRIPTION
**Pull request by theflockers**
This PR adds the missing metrics for the release-service to implement the alert/visualisations for the agreed SLO

### CODE REVIEW

File: metrics/release.go [hunk 1]
```diff
@@ -208,6 +236,39 @@ func RegisterCompletedReleaseProcessing(startTime, completionTime *metav1.Time,
...
+
+	ReleaseValidationSeconds.
+		With(prometheus.Labels{
+			"reason": reason,
+			"target": target,
+		}).
+		Observe(validationTime.Sub(startTime.Time).Seconds())
+}
+
+// RegisterStartedReleaseProcessing registers a Release as start processing, adding a new observation for the
```
Comment (johnbieren) [inline metrics/release.go:255]: s/start/started/

File: metrics/release.go [hunk 2]
```diff
@@ -208,6 +236,39 @@ func RegisterCompletedReleaseProcessing(startTime, completionTime *metav1.Time,
...
+		return
+	}
+
+	ReleaseStartProcessingSeconds.
+		With(prometheus.Labels{
+			"reason": reason,
+			"target": target,
+		}).
+		Observe(processingStartTime.Sub(startTime.Time).Seconds())
+	ReleaseConcurrentProcessingsTotal.WithLabelValues().Dec()
```
Comment (johnbieren) [inline metrics/release.go:269]: I'm not sure we want to decrement the total number of releases processing here

Comment (theflockers) [reply metrics/release.go]: removed.

**Review by johnbieren — APPROVED** [review]
LGTM, but I think it would be nice to give David a chance to look it over before merging

File: README.md [hunk 3]
```diff
@@ -67,4 +67,6 @@ by default, this operator exports the following custom metrics:
 | release_duration_seconds                         | Histogram | How long in seconds a Release takes to complete.                    |
 | release_post_actions_execution_duration_seconds  | Histogram | How long in seconds Release post-actions take to complete.          |
 | release_processing_duration_seconds              | Histogram | How long in seconds a Release processing takes to complete.         |
+| release_start_processing_seconds                 | Histogram | How long in seconds a Release takes to start processing             |
+| release_validation_seconds                       | Histogram | How long in seconds a Release takes to validate                     |
```
Comment (davidmogar) [inline README.md:71]: Following the other names, this metric should be called `release_validation_duration_seconds`.

File: README.md [hunk 4]
```diff
@@ -67,4 +67,6 @@ by default, this operator exports the following custom metrics:
 | release_duration_seconds                         | Histogram | How long in seconds a Release takes to complete.                    |
 | release_post_actions_execution_duration_seconds  | Histogram | How long in seconds Release post-actions take to complete.          |
 | release_processing_duration_seconds              | Histogram | How long in seconds a Release processing takes to complete.         |
+| release_start_processing_seconds                 | Histogram | How long in seconds a Release takes to start processing             |
```
Comment (davidmogar) [inline README.md:70]: What about `release_pre_processing_duration_seconds`?

File: metrics/release.go [hunk 5]
```diff
@@ -208,6 +236,38 @@ func RegisterCompletedReleaseProcessing(startTime, completionTime *metav1.Time,
...
+			"reason": reason,
+			"target": target,
+		}).
+		Observe(validationTime.Sub(startTime.Time).Seconds())
+}
+
+// RegisterStartedReleaseProcessing registers a Release as started processing, adding a new observation for the
+// Release start processing duration and increasing the number of concurrent processings. If either the
+// processingStartTime or the releaseStartTime are nil, no action will be taken.
+func RegisterStartedReleaseProcessing(startTime, processingStartTime *metav1.Time, reason, target string) {
```
Comment (davidmogar) [inline metrics/release.go:258]: You don't need a new function I think. You should use the `RegisterNewReleaseProcessing` as that's the moment where the phase changes.

File: api/v1alpha1/release_types.go [hunk 6]
```diff
@@ -331,7 +332,12 @@ func (r *Release) MarkProcessing(message string) {
 
 	conditions.SetConditionWithMessage(&r.Status.Conditions, processedConditionType, metav1.ConditionFalse, ProgressingReason, message)
 
-	go metrics.RegisterNewReleaseProcessing()
```
Comment (davidmogar) [inline api/v1alpha1/release_types.go:334]: Why are you removing this?

File: README.md (hunk 3 above)
Comment (theflockers) [reply README.md]: changed.

File: README.md (hunk 4 above)
Comment (theflockers) [reply README.md]: changed.

File: metrics/release.go (hunk 5 above)
Comment (theflockers) [reply metrics/release.go]: Function removed. I moved the content of it to the `RegisterNewReleaseProcessing`.

File: api/v1alpha1/release_types.go (hunk 6 above)
Comment (theflockers) [reply api/v1alpha1/release_types.go:334]: added back, but now with parameters.

File: metrics/release.go [hunk 7]
```diff
@@ -218,8 +262,20 @@ func RegisterNewReleaseDeployment() {
...
 
-// RegisterNewReleaseProcessing register a new Release processing, increasing the number of concurrent processings.
-func RegisterNewReleaseProcessing() {
+// RegisterNewReleaseProcessing registers a new Release processing, adding a new observation for the
+// Release start processing duration and increasing the number of concurrent processings. If either the
+// processingStartTime or the releaseStartTime are nil, no action will be taken.
+func RegisterNewReleaseProcessing(startTime, processingStartTime *metav1.Time, reason, target string) {
+	if startTime == nil || processingStartTime == nil {
+		return
+	}
```
Comment (davidmogar) [inline metrics/release.go:271]: Could you add a line break like in the other cases?

File: metrics/release.go [hunk 8]
```diff
@@ -218,8 +262,20 @@ func RegisterNewReleaseDeployment() {
 	ReleaseConcurrentDeploymentsTotal.WithLabelValues().Inc()
 }
 
-// RegisterNewReleaseProcessing register a new Release processing, increasing the number of concurrent processings.
-func RegisterNewReleaseProcessing() {
+// RegisterNewReleaseProcessing registers a new Release processing, adding a new observation for the
+// Release start processing duration and increasing the number of concurrent processings. If either the
+// processingStartTime or the releaseStartTime are nil, no action will be taken.
```
Comment (davidmogar) [inline metrics/release.go:267]: There's no releaseStartTime.

File: metrics/release.go (hunk 7 above)
Comment (theflockers) [reply metrics/release.go:271]: done

File: metrics/release.go (hunk 8 above)
Comment (theflockers) [reply metrics/release.go]: done

**Review by davidmogar — APPROVED** [review]

**Pull request merged**